helm-charts/splunk-otel-collector/templates/config/_otel-eks-fargate-cluster-receiver-observer.tpl

```diff
@@ -0,0 +1,205 @@
+{{/*
+Config for the otel-collector eks/fargate cluster receiver observer deployment.
+The values can be overridden in .Values.clusterReceiverObserver.config
+*/}}
+{{- define "splunk-otel-collector.clusterReceiverObserverConfig" -}}
+{{ $gateway := fromYaml (include "splunk-otel-collector.gateway" .) -}}
+{{ $clusterReceiverObserver := fromYaml (include "splunk-otel-collector.clusterReceiverObserver" .) -}}
+extensions:
+  health_check:
+
+  memory_ballast:
+    size_mib: ${SPLUNK_BALLAST_SIZE_MIB}
+
+  # k8s_observer w/ pod and node detection for eks/fargate deployment
+  k8s_observer:
+    auth_type: serviceAccount
+    observe_pods: true
+    observe_nodes: true
+
+receivers:
+  # Prometheus receiver scraping metrics from the pod itself
+  prometheus/k8s_cluster_receiver:
+    config:
+      scrape_configs:
+      - job_name: 'otel-k8s-cluster-receiver-observer'
+        scrape_interval: 10s
+        static_configs:
+        - targets: ["${K8S_POD_IP}:8889"]
+  {{- if $clusterReceiverObserver.k8sEventsEnabled }}
+  smartagent/kubernetes-events:
+    type: kubernetes-events
+    alwaysClusterReporter: true
+    whitelistedEvents:
+    - reason: Created
+      involvedObjectKind: Pod
+    - reason: Unhealthy
+      involvedObjectKind: Pod
+    - reason: Failed
+      involvedObjectKind: Pod
+    - reason: FailedCreate
+      involvedObjectKind: Job
+  {{- end }}
+
+  # dynamically created kubeletstats receiver to report kubelet stats for cluster receiver "node"
+  receiver_creator/eks-fargate-cluster-receiver:
+    receivers:
+      kubeletstats:
+        rule: type == "k8s.node" && name contains "fargate" && labels["otel-eks-fargate-is-cluster-receiver-node"] == "true"
```

Comment:
Isn't this rule checking for node labels while the label we are looking for is set on a pod?

Reply:
yikes, good catch.

Reply:
@dmitryax I'm not sure of a way that helm can manage this and fargate profiles don't appear to be able add node labels either (only filter from pod labels). Other than requiring users to manually set a label w/ kubectl or similar can you think of an approach?

Reply:
related that suggests adequate message passing though pod labels isn't possible: https://github.com/kubernetes/kubernetes/issues/40610

Reply:
I was able to resolve this with an init container on the cluster receiver running kubectl but it requires a patch role for nodes, which is likely undesirable.

Reply:
It's weird that they allow patching nodes in Fargate :) It does seem undesirable and likely can be disabled at some point. If we are going this route we need to make sure that it's possible to patch all the nodes to remove stale labels.

Do they allow running pods in hostNetwork? If so, maybe we can add another static receiver talking to localhost?

Reply:
Fargate nodes are ephemeral and are only around as long as they have a pod so cleanup isn't necessary tmk. `Pods running on Fargate can't specify HostPort or HostNetwork in the pod manifest` (confirmed they won't start).

Reply:
Does it mean that pods in Fargate are always scheduled on dedicated nodes, and the same node cannot have more than one pod? Based on your solution, looks like it's assumed.

Assuming ^ I we can consider another option:

Reply:
Does it mean that pods in Fargate are always scheduled on dedicated nodes, and the same node cannot have more than one pod? Based on your solution, looks like it's assumed.

I'm also thinking of the another option... if fargate is distro is enabled we:
1. Change clusterReceiver to 2-replicas statefulSet.
2. Add an initContainer that don't patch the node, but prepares the discovery rule config line that can be injected in the config using some config source (from file to env var?). To get the discovery rule it checks which replica it's running on: 
 - the first replica gets the known line.
 - the second replica looks for first replica node's IP address using the k8s API and prepares the discovery rule using that IP address.

What do you think? Can it work?

Reply:
Also if the If it's not the case that each Fargate node can have only one pod, we need to setup podAntiAffinity for the statefulSet ^

Reply:
I'm not seeing any assurances that each fargate node is unique, but in practice this has been what I've observed. Thanks for the statefulSet suggestion, which I'm looking into now.

Reply:
Also we need to remove `k8s_cluster` receiver from the pipeline of one of the replicas, everything else in the configs should be identical I believe. As an option we can copy configmap's config file to another local volume shared between containers and adjust it accordingly in the init container with a tool like https://mikefarah.gitbook.io/yq/.

Reply:
@dmitryax after getting up to speed and [trying this out (ignore the readmes)](https://github.com/signalfx/splunk-otel-collector-chart/commit/663619a864149e775bf51d9aa3dbc423b7ca52bb) I don't think this approach works well:

1. Because they are launched sequentially pod-0 can only know not to watch it itself by filtering its node-0 name from the downward api in the kubelet stats receiver (and not have a cluster receiver). pod-1 knows to watch only node-0's kubelet by querying pod-0 (and have the cluster receiver) in an init container (read: static config).
2. When pod-0/node-0 goes down there is no updating the filter in pod-1 so monitoring pod-0/node-0 is permanently broken and self-monitoring doesn't work, defeating the whole point.

My proposal is now to keep the current node patching logic but have a 2-replica stateful set w/ node labelling init containers to keep the messaging passing dynamic and the collector instances linked. This way when pod-0/node-0 goes down it will reregister itself via node label so that pod-1 tracks the node change.  I think this also gives us the benefit of being able to use parallel pod management since they work independently but still have significant name ordinals (no waiting w/ intermittent cluster reporting).

Reply:
Yes this is a problem. I think there should be a way to enforce restart of the second pod based on the liveness of the first one. I was under impression that this dependency somehow enforced not only during the first start.

First crazy idea that crosses my mind is having a regular sidecar container on the second pod that would watch for the first pod and kill itself if the first pod is unavailable. Or maybe [custom pod readiness gate](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-readiness-gate) maybe can help. 

I'm not saying we should try anything like this right now. I think we can proceed with what you just suggested (statefulset + patching node) and maybe later come up with another solution to avoid node pathing.